File: DSA/src/Questions/BinarySearch/FindKClosestElement.java
package Questions.BinarySearch;

import java.util.ArrayList;
import java.util.List;
/*
https://leetcode.com/problems/find-k-closest-elements/description/
 Given a sorted integer array arr, two integers k and x, return the k closest integers to x in the array. The result should also be sorted in ascending order.

An integer a is closer to x than an integer b if:

|a - x| < |b - x|, or
|a - x| == |b - x| and a < b
 

Example 1:

Input: arr = [1,2,3,4,5], k = 4, x = 3
Output: [1,2,3,4]
Example 2:

Input: arr = [1,2,3,4,5], k = 4, x = -1
Output: [1,2,3,4]
 

Constraints:

1 <= k <= arr.length
1 <= arr.length <= 104
arr is sorted in ascending order.
-104 <= arr[i], x <= 104
Time Complexity: O(log(N) + k)
Space Complexity: O(1)
 */
public class FindKClosestElement {
public static List<Integer> findClosestElements(int[] arr, int k, int x) {
        int left = 0, right = arr.length - k;

        // Binary search to find the start of the window
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (x - arr[mid] > arr[mid + k] - x) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        List<Integer> result = new ArrayList<>();
        for (int i = left; i < left + k; i++) {
            result.add(arr[i]);
        }

        // Since the elements are already in the correct order
        return result;
    }
    public static void main(String[] args) {
        // Example usage
        int[] arr = {1, 2, 3, 4, 5};
        int k = 4, x = 3;
        System.out.println(findClosestElements(arr, k, x)); // Expected output: [1, 2, 3, 4]
    }
}

File: DSA/Python/Questions/BinarySearch/FindKClosestElement.py
"""
# Question: Find K Closest Elements
# Link: https://leetcode.com/problems/find-k-closest-elements/

# Problem Statement:
# Given a sorted integer array arr, two integers k and x, return the k closest integers to x in the array.
# The result should also be sorted in ascending order.

# Example:
# Input: arr = [1,2,3,4,5], k = 4, x = 3
# Output: [1,2,3,4]
# Explanation: The 4 closest elements to 3 are [1,2,3,4]
"""

from typing import List
import heapq


class FindKClosestElement:
    def find_closest_binary_search(self, arr: List[int], k: int, x: int) -> List[int]:
        """
        Binary search approach to find starting position
        Time Complexity: O(log(n-k))
        Space Complexity: O(1)
        """
        left, right = 0, len(arr) - k

        while left < right:
            mid = (left + right) >> 1
            if x - arr[mid] > arr[mid + k] - x:
                left = mid + 1
            else:
                right = mid

        return arr[left : left + k]

    def find_closest_heap(self, arr: List[int], k: int, x: int) -> List[int]:
        """
        Heap-based approach using absolute differences
        Time Complexity: O(nlogk)
        Space Complexity: O(k)
        """
        heap = []
        for num in arr:
            diff = abs(num - x)
            heapq.heappush(heap, (-diff, -num))
            if len(heap) > k:
                heapq.heappop(heap)

        result = [-pair[1] for pair in heap]
        return sorted(result)

    def find_closest_two_pointer(self, arr: List[int], k: int, x: int) -> List[int]:
        """
        Two pointer approach
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        left = 0
        right = len(arr) - 1

        while right - left >= k:
            if abs(arr[left] - x) > abs(arr[right] - x):
                left += 1
            else:
                right -= 1

        return arr[left : right + 1]


def main():
    test_cases = [
        {"arr": [1, 2, 3, 4, 5], "k": 4, "x": 3, "expected": [1, 2, 3, 4]},
        {"arr": [1, 2, 3, 4, 5], "k": 4, "x": -1, "expected": [1, 2, 3, 4]},
        {"arr": [1, 1, 1, 10, 10, 10], "k": 1, "x": 9, "expected": [10]},
    ]

    solution = FindKClosestElement()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Array: {test['arr']}")
        print(f"k: {test['k']}, x: {test['x']}")

        result_binary = solution.find_closest_binary_search(
            test["arr"], test["k"], test["x"]
        )
        result_heap = solution.find_closest_heap(test["arr"], test["k"], test["x"])
        result_two_pointer = solution.find_closest_two_pointer(
            test["arr"], test["k"], test["x"]
        )

        print(f"Binary Search Result: {result_binary}")
        print(f"Heap Result: {result_heap}")
        print(f"Two Pointer Result: {result_two_pointer}")
        print(f"Expected: {test['expected']}")

        assert result_binary == result_heap == result_two_pointer == test["expected"]


if __name__ == "__main__":
    main()
